Template Api Support

pwittrock opened the pull request:
``` release-note
* Add extensions/v1beta1 'Template' Api object to support parameter substitution into configs
```

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/25293)

<!-- Reviewable:end -->

pwittrock commented:
cc @bparees Since you worked on the original design proposal it might be helpful for you to take a look at the api

cc @kubernetes/kube-api

pwittrock commented:
All commits including "fixup" and earlier are from another PR that this depends on.  No need to review them.

bparees commented on pkg/apis/extensions/types.go:
```diff
@@ -818,3 +819,80 @@ type PodSecurityPolicyList struct {
+
+	// Optional: Value holds the Parameter data. If specified, the generator
+	// will be ignored. The value replaces all occurrences of the Parameter
+	// ${Name} expression during the Template to Config transformation.
```
there will no generators in the k8s implementation, and the substitution string/syntax will be $(Name) instead of ${Name}  (actually the syntax will be $(Name) and $((Name)) depending whether you want the substituted result to be quoted or not)

pwittrock replied:
Done

bparees commented:
> cc @bparees Since you worked on the original design proposal it might be helpful for you to take a look at the api

the api objects look good, i'm a little fuzzier on the rest api, particularly around processing templates.  In the spec we talked about 2 key aspects:
1) whether you post parameters, or a full template
2) whether you get back a full template(with substitution complete), or a list of objects

it looks like what's defined here is:  post parameters, get back a list of objects.

I think we may still want the "post a template" input option so users can process a template without having to first "create" it.  (if that's there and i missed it, sorry)

for the response object, i think just returning the list of objects (instead of the full template) is right.

kargakis commented:
cc: @mfojtik

pwittrock commented:
>
I think we may still want the "post a template" input option so users can process a template without having to first "create" it. (if that's there and i missed it, sorry)

Sounds good.  I will talk with @lavalamp about how to best expose this in api.  Do you have a preference (e.g. different process endpoints for different types, a single request that wraps both input types, etc)?

bparees commented:
> Sounds good. I will talk with @lavalamp about how to best expose this in api. Do you have a preference (e.g. different process endpoints for different types, a single request that wraps both input types, etc)?

a separate endpoint makes sense to me...  logically you'd then have:
1) an api/\<namespace\>/templates/\<templatename\>/process api that you can post parameters to, which will be combined with that specific template
and
2) an api/\<namespace(?)\>/processtemplate api that takes a template input for processing.  (having a namespace in the path doesn't make much sense here since you're not actually interacting with the namespace)

pwittrock commented:
@bparees Daniel agreed the second endpoint it sensible way of doing this.  Since this is the second to last week of coding for 1.3 I'd like to minimize the scope to an MVP before addressing anything else.  Would you be ok with me adding a TODO for adding the second api endpoint once the first endpoint is complete?

bparees commented:
> Would you be ok with me adding a TODO for adding the second api endpoint once the first endpoint is complete?

sounds reasonable

lavalamp commented:
Making the processtemplate endpoint cluster-scoped (no namespace) would make sense as long as there's no fan-out (calls to other resources).

Maybe too late to change now, but "config" might be a good group name for both template resources.

pwittrock commented:
@lavalamp How hard would it be to create a config/v1/beta1 api group?  Do you think it is worth it / Does it put this feature at risk for 1.3?

lavalamp commented:
It requires a bunch of counter-intuitive boilerplate, basically, although
you can mostly copy-pasta. Maybe not worth it this late in the process.

On Mon, May 9, 2016 at 2:06 PM, Phillip Wittrock <notifications@github.com>
wrote:

> @lavalamp <https://github.com/lavalamp> How hard would it be to create a
> config/v1/beta1 api group? Do you think it is worth it / Does it put this
> feature at risk for 1.3?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/25293#issuecomment-217989714>
>

pwittrock commented:
Leaving in extensions SGTM then.  Just getting the api hooked up was mainly copy paste, but it took me longer than expected to figure out exactly how it needed to be done.

pwittrock commented:
PTAL

janetkuo commented on pkg/apis/extensions/types.go:
```diff
@@ -818,3 +819,82 @@ type PodSecurityPolicyList struct {
+	Parameters []Parameter `json:"parameters,omitempty"`
+
+	// Required: A yaml config list of resources to create
+	Items []runtime.RawExtension `json:"items,omitempty"`
```
We should use `Object` in internal struct, right?

janetkuo replied:
Thinking out loud: `Items` sound like a slice of object xxx of xxxList (e.g. []Pod of PodList).

pwittrock replied:
runtime.Object won't work because trying to parse the items into actual go objects will fail.  (Trying to parse TEMPLATE_PARAM_NAME into an int field would fail)

pwittrock replied:
> Thinking out loud: Items sound like a slice of object xxx in xxxList (e.g. []Pod in PodList).

That was intentionally done to match the output result (a List).  WDYT?

janetkuo replied:
Hmm but [`runtime.RawExtension` doc](https://github.com/kubernetes/kubernetes/blob/master/pkg/runtime/types.go#L48-L99) stated:

> 
```go
// To use this, make a field which has RawExtension as its type in your external, versioned
// struct, and Object in your internal struct. You also need to register your
// various plugin types.
```

janetkuo replied:
Looking at other struct in types.go it seems that `Items` field is only used as the items of a List type struct. Not sure if it's convention. Suggest naming it `Objects` or `Resources`.

smarterclayton replied:
We call it objects in Openshift - I would favor that.  We did not call it items because we reserve that for true lists

pwittrock replied:
SGTM, I can switch it back to Objects

janetkuo commented on pkg/client/unversioned/templates.go:
```diff
@@ -0,0 +1,95 @@
+	Watch(opts api.ListOptions) (watch.Interface, error)
+}
+
+var _ TemplateInterface = &templates{}
```
Suggest adding a comment here:
```go
// Ensure statically that templates implements TemplateInterface.
```

janetkuo commented on pkg/client/unversioned/templates.go:
```diff
@@ -0,0 +1,95 @@
+	ns     string
+}
+
+// newTemplates returns a Templates
```
a templates

pwittrock replied:
Done

janetkuo commented on pkg/client/unversioned/testclient/fake_templates.go:
```diff
@@ -0,0 +1,85 @@
+	Fake      *FakeExperimental
+	Namespace string
+}
+
```
Add a static check to make sure FakeTemplates implements TemplatesInterface here too

pwittrock replied:
Done

janetkuo commented:
Actually we don't need to write clients ourselves. Should use generated clients instead: see https://github.com/kubernetes/kubernetes/blob/master/docs/devel/generating-clientset.md#using-client-gen, adding `// +genclient=true` tag in your types.go will create generated clients for you.

pwittrock committed:
Add Template Api types and wiring

Feature Issue: kubernetes/features#3

- Add Template types to extensions/types.go and extensions/v1beta1/types.go
- Add etcd storage for Template object
- Add stragegy for Template object
- Register Template Api types in the extensions api
- Register Template Api CRUD and Process endpoints in the master
- Add generated client expansions
- Add kubectl 'get' and 'describe' support for Template Api objects

pwittrock committed:
Fix serialization issue with fuzzer for raw Object types

Stip the trailing newline from the raw bytes.

pwittrock commented:
PTAL

janetkuo commented:
Added more comments in https://github.com/kubernetes/kubernetes/pull/25293#discussion-diff-62754866

janetkuo commented on pkg/kubectl/cmd/cmd.go:
```diff
@@ -171,6 +171,7 @@ __custom_func() {
    * resourcequotas (aka 'quota')
    * replicasets (aka 'rs')
    * replicationcontrollers (aka 'rc')
+   * templates
```
This is used by `get` and `describe`, so you'll need to implement its describer too.

janetkuo replied:
I meant template's

pwittrock replied:
Done

janetkuo commented on pkg/registry/template/etcd/etcd_test.go:
```diff
@@ -0,0 +1,130 @@
+			object := obj.(*extensions.Template)
+			object.Spec.ObjectLabels = map[string]string{"c": "d"}
+			return object
+		},
```
Suggest add several invalid updates here. However since template validation isn't implemented yet, add TODO here and mention this in validation.go also.

pwittrock replied:
Done

pwittrock commented:
@k8s-bot integration test this

janetkuo commented on pkg/kubectl/describe.go:
```diff
@@ -1231,6 +1233,44 @@ func describeSecret(secret *api.Secret) (string, error) {
+		fmt.Fprintf(out, "Labels (Count: %v):\n    Key\tValue\n", len(temp.Spec.ObjectLabels))
+		fmt.Fprint(out, "    ---\t-----\n")
+		for k, v := range temp.Spec.ObjectLabels {
+			fmt.Fprintf(out, "    %s\t%s\n", k, v)
```
You can use `labels.FormatLabels(temp.Spec.ObjectLabels)` to print this as a single line, or use `printLabelsMultilineWithIndent` to print it as multi-lines (with "    " indent).

janetkuo commented:
Looks good now. @bgrant0607 might want to take a look on api changes.

pwittrock commented:
@k8s-bot unit test this

bgrant0607 commented on pkg/apis/extensions/v1beta1/types.go:
```diff
@@ -1102,3 +1103,87 @@ type PodSecurityPolicyList struct {
+	Parameters []Parameter `json:"parameters,omitempty" protobuf:"bytes,2,rep,name=parameters"`
+
+	// Required: A yaml config list of resources to create
+	Objects []runtime.RawExtension `json:"objects,omitempty" protobuf:"bytes,3,opt,name=objects"`
```
Required fields shouldn't have omitempty

pwittrock replied:
Done.

bgrant0607 commented on pkg/apis/extensions/v1beta1/types.go:
```diff
@@ -1102,3 +1103,87 @@ type PodSecurityPolicyList struct {
+	Value string `json:"value,omitempty" protobuf:"bytes,4,opt,name=value"`
+
+	// Optional: Indicates the parameter must have a value.  Defaults to false.
+	Required bool `json:"required,omitempty" protobuf:"varint,7,opt,name=required"`
```
Please make this a pointer so we can distinguish unspecified from false in kubectl apply.

pwittrock replied:
Shouldn't all Optional params be pointers then?

pwittrock replied:
Updated all the Optional fields to be Pointers per the api-conventions doc.

bgrant0607 commented on pkg/apis/extensions/v1beta1/types.go:
```diff
@@ -1102,3 +1103,87 @@ type PodSecurityPolicyList struct {
+type TemplateParameters struct {
+	unversioned.TypeMeta `json:",inline"`
+	v1.ObjectMeta        `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	Spec                 TemplateParametersSpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
```
Doesn't need `spec`

bgrant0607 commented on pkg/apis/extensions/v1beta1/types.go:
```diff
@@ -1102,3 +1103,87 @@ type PodSecurityPolicyList struct {
+}
+
+type TemplateParametersSpec struct {
+	TemplateName string `json:"templateName" protobuf:"bytes,3,opt,name=templateName"`
```
I'd just call it `Name`.

pwittrock replied:
Done

bgrant0607 commented on pkg/apis/extensions/v1beta1/types.go:
```diff
@@ -1102,3 +1103,87 @@ type PodSecurityPolicyList struct {
+// TemplateParameters contains the substitution parameters for processing a Template
+type TemplateParameters struct {
+	unversioned.TypeMeta `json:",inline"`
+	v1.ObjectMeta        `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
```
Is this a namespace-level resource, or a subresource of the template?

bgrant0607 replied:
Or do we want to use the same type for both?

pwittrock replied:
This is a subresource of the template.  Per our conversation Friday, I updated it to contain just a map[string]string instead of the Parameter structs.  I also removed ObjectMeta and just added a 'Name' field (matches DeploymentRollback)

smarterclayton replied:
We used the same type for both - since we already have APIs that are POST
-> get back same object filled out (subjectaccessreview)

On Fri, May 13, 2016 at 7:23 PM, Brian Grant <notifications@github.com>
wrote:

> In pkg/apis/extensions/v1beta1/types.go
> <https://github.com/kubernetes/kubernetes/pull/25293#discussion_r63263697>
> :
>
> > +	// The value replaces all occurrences of the Parameter $(Name) or
> > +	// $((Name)) expression during the Template to Config transformation.
> > +	Value string `json:"value,omitempty" protobuf:"bytes,4,opt,name=value"`
> > +
> > +	// Optional: Indicates the parameter must have a value.  Defaults to false.
> > +	Required bool `json:"required,omitempty" protobuf:"varint,7,opt,name=required"`
> > +
> > +	// Optional: Type-value of the parameter (one of string, int, bool, or base64)
> > +	// Used by clients to provide validation of user input and guide users.
> > +	Type ParameterType `json:"type,omitempty" protobuf:"bytes,8,opt,name=type,casttype=ParameterType"`
> > +}
> > +
> > +// TemplateParameters contains the substitution parameters for processing a Template
> > +type TemplateParameters struct {
> > +	unversioned.TypeMeta `json:",inline"`
> > +	v1.ObjectMeta        `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
>
> Or do we want to use the same type for both?
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/25293/files/38c30b66714dfc3fef1f7872e45652db8273980f#r63263697>
>

bgrant0607 commented:
Re. optional params:
https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api-conventions.md#optional-vs-required

pwittrock commented:
@k8s-bot test this issue: #DROPPED

bparees commented on pkg/apis/extensions/types.go:
```diff
@@ -809,3 +810,75 @@ type PodSecurityPolicyList struct {
+	unversioned.TypeMeta `json:",inline"`
+
+	// This must match the Name of a template
+	Name string `json:"name"`
```
what's the intended purpose of this field?

pwittrock replied:
It is the name of the template to process as well as the name of the parameters.

lavalamp replied:
You shouldn't need this. The name will appear in the URL path.

pwittrock replied:
I believe we will need this to support kubectl -f params.yaml without specifying the name on the commandline.  Thoughts?

lavalamp replied:
Hm. You can make a wrapping type for that, I guess. But if you add it here it'll be weird because it'll be in the URL path and here also.

bgrant0607 replied:
The Create handler is passed just the context and the object. 

Other RESTy subresources explicitly contain the name. Binding and Scale include ObjectMeta. DeploymentRollback (#19686) contains just the Name. 

Namespace is available in api.Context, but I don't think Name is?

bparees replied:
> I believe we will need this to support kubectl -f params.yaml 

is that "kubectl create -f params.yaml"?

is create via a parameters object leading to an instantiated template a path we actually want to support?  What is actually created?  The template processing APIs themselves just return content, so this create invocation would not actually modify etcd, which seems unusual.

are there other cases in k8s where a "create" from a resource file does not actually create that resource in etcd?

smarterclayton replied:
It is not but @derekwaynecarr and I were discussing passing an
ObjectMeta all the way down through the context eventually.

lavalamp replied:
I hope that is not true, it'd mean you could pass a binding for pod foo to
pods/bar...
On May 16, 2016 5:10 PM, "Brian Grant" <notifications@github.com> wrote:

> In pkg/apis/extensions/types.go
> <https://github.com/kubernetes/kubernetes/pull/25293#discussion_r63447506>
> :
>
> > +	Value string `json:"value,omitempty"`
> > +
> > +	// Indicates the parameter must have a value.
> > +	Required bool `json:"required,omitempty"`
> > +
> > +	// Type-value of the parameter (one of string, int, bool, or base64)
> > +	// Used by clients to provide validation of user input and guide users.
> > +	Type string `json:"type,omitempty"`
> > +}
> > +
> > +// TemplateParameters contains the substitution parameters for processing a Template
> > +type TemplateParameters struct {
> > +	unversioned.TypeMeta `json:",inline"`
> > +
> > +	// This must match the Name of a template
> > +	Name string `json:"name"`
>
> The Create handler is passed just the context and the object.
>
> Other RESTy subresources explicitly contain the name. Binding and Scale
> include ObjectMeta. DeploymentRollback (#19686
> <https://github.com/kubernetes/kubernetes/pull/19686>) contains just the
> Name.
>
> Namespace is available in api.Context, but I don't think Name is?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/25293/files/eb8d3942af3029e3c15d78d29faa4d14b419edad#r63447506>
>

pwittrock replied:
Makes sense.  Lets talk about how we can solve this tomorrow.  I agree it doesn't make sense in the swagger spec since this is a sub-resource of template.

pwittrock replied:
@bparees it would be `kubectl process template -f params.yaml`

bparees replied:
>  @bparees it would be kubectl process template -f params.yaml

so why not `kubectl process template/\<templatename\> -f params.yaml` to avoid having to include the name in the parameters?

(sorry if i'm suggesting things counter to k8s patterns, i don't work in k8s much)

pwittrock replied:
That seems like a reasonable way to go.  Passing multiple `-f` args would get a bit weird, but we can just prohibit that.

bparees commented on pkg/apis/extensions/types.go:
```diff
@@ -809,3 +810,75 @@ type PodSecurityPolicyList struct {
+
+	// Required: A yaml config list of resources to create
+	Objects []runtime.RawExtension `json:"objects"`
+}
```
did we lose the ObjectLabels field (labels to be applied to all objects created from the template)?  if so, why?

pwittrock replied:
Yes.  Apologies for not looping you into this directly.  I chatted with @lavalamp and @bgrant0607 a bit about this and decided to defer ObjectLabels support.

There were 2 main concerns with how ObjectLabels should behave: 
- Objects such as Deployment use label defaulting to apply pod-labels to the Deployment if the Deployment has no labels.  ObjectLabels would break the defaulting because the Deployment labels would be non-empty
- What is the correct thing to do with recursion?  Should the labels be applied to the PodTemplate object in a Deployment?  What about elements of Lists?

We may want to instead consider supporting map[string]string as a substitution type so that the user can be more explicit about how they want things wired together.

Since it will be easier to add some support for object labels in the future than it will be to change the behavior if we already support them, I think we should hold off on this until we know exactly what we want.

pwittrock replied:
@bparees If you have time, it might be helpful in chatting over VC to talk about of a few related changes.

bparees replied:
Unfortunately I'm traveling this week, not sure I'll have a good opportunity.

bgrant0607 replied:
@bparees Yes, we're punting on that for now, in the spirit of getting something done. Specific issues mentioned were:
* breaking defaulting of controller labels from pod template
* not propagating labels to pod template

bparees replied:
got it.  yeah we've had some issues w/ the question of what to propagate to pod templates too. currently we only apply the labels to top level objects, so we do not apply them to pod templates within a deployment config, for example.  but that has definitely caused some end-user confusion.

pwittrock replied:
@bparees No problem.  Another change I made was to only support $(NAME) and not $((NAME)).  I did this because the Template processing is performed on the server side and the json / yaml has already been parsed.  I restricted parameters names to be [A-Z0-9_]+ since that was the naming convention used in the design doc.

bparees replied:
>  Another change I made was to only support $(NAME) and not $((NAME)). I did this because the Template processing is performed on the server side and the json / yaml has already been parsed.

@pwittrock i'm not following the connection between those two statements.  Is the lack of support for $((NAME)) a point in time statement?  Or is there another plan for how to support allowing substitution of non-string fields?

pwittrock replied:
Non-string fields will be supported.  I need to better understand why we need $((NAME))  vs $(NAME) to support non-string fields.  Since we are unmarshaling the json into golang structs on the server side, is it sufficient to use the type of the field we are unmarshaling into to determine how we parse the field?  Given Parameter "TRUE" substituted with "true" can we say for string fields we will parse and assign the `string "true"` and for boolean fields we will parse and assign the `bool true`?

smarterclayton replied:
There are places where the parameter would be used as both string and
number - for instance, REPLICAS used on both the "replicas" field of a
PetSet and as the value of environment variable REPLICA_COUNT. There
is no guarantee the server knows (or can know) the fields of the
objects stored in a template.  Processing of the template needs to be
done without casting to a go type, because it should be possible to
store a template on a server for which some fields are not known to
that server without losing those fields (fields added later to API
v1).

pwittrock commented:
@k8s-bot e2e node test this

pwittrock commented:
PTAL

pwittrock commented:
@k8s-bot node e2e test this

bgrant0607 commented on pkg/apis/extensions/v1beta1/types.go:
```diff
@@ -1102,3 +1103,78 @@ type PodSecurityPolicyList struct {
+}
+
+type TemplateSpec struct {
+	// Optional: Parameters is an array of Parameters used during the
```
These comments appear in user-facing documentation. Please use json field names and plain English to describe fields and types, or don't mention the field name at all. 

Also, we usually don't specify optional vs. required in the comment.That's indicated by the presence or absence of `omitempty`.

pwittrock replied:
Will update

bgrant0607 commented on pkg/apis/extensions/v1beta1/types.go:
```diff
@@ -1102,3 +1103,78 @@ type PodSecurityPolicyList struct {
+	// Template to Config transformation.
+	Parameters []Parameter `json:"parameters,omitempty" protobuf:"bytes,2,rep,name=parameters"`
+
+	// Required: A yaml config list of resources to create
```
Does it have to be yaml? Could it be json?

pwittrock replied:
Yup json is fine.  Will update

pwittrock replied:
Thinking on this further, this may not be required to say at all.  The Objects are just Objects.  It is a bit weird since here they are RawExtensions and have a '[]bytes raw' field, but I think it is clear enough.

bgrant0607 commented on pkg/apis/extensions/v1beta1/types.go:
```diff
@@ -1102,3 +1103,78 @@ type PodSecurityPolicyList struct {
+	// Defaults to false.
+	Required *bool `json:"required,omitempty" protobuf:"varint,7,opt,name=required"`
+
+	// Optional: Type-value of the parameter (one of string, int, bool, or base64)
```
These type names don't match the constants above.

pwittrock replied:
Will update.

bgrant0607 commented:
@pwittrock Could you please squash into one commit with manual changes and another with all the generated files?

pwittrock committed:
Add Template Api generated code

Feature Issue: kubernetes/features#3

pwittrock commented:
Squashed

pwittrock commented:
The validation and processing logic will be in #25622

pwittrock commented:
@k8s-bot node e2e test this issue: #25735

pwittrock commented:
PTAL.

lavalamp commented on pkg/apis/extensions/types.go:
```diff
@@ -809,3 +810,80 @@ type PodSecurityPolicyList struct {
+	unversioned.TypeMeta `json:",inline"`
+
+	// Name is the name of the Template to be processed.
+	Name string `json:"name"`
```
Looks like Brian is correct. Consider just using object meta instead, to be consistent with Binding.

pwittrock replied:
ObjectMeta has a bunch of stuff like namespace and nameGenerator that I don't think we want.  This is consistent with extensions/v1beta.DeploymentRollback and I think I prefer it.  WDYT?

smarterclayton replied:
We use ObjectMeta in many places for virtual resources (like
SubjectAccessReview).

On Tue, May 17, 2016 at 2:53 PM, Daniel Smith <notifications@github.com>
wrote:

> In pkg/apis/extensions/types.go
> <https://github.com/kubernetes/kubernetes/pull/25293#discussion_r63581936>
> :
>
> > +	Required bool `json:"required,omitempty"`
> > +
> > +	// Type is the type that the parameter value must be parsed to. Type may be one of
> > +	// 'string', 'integer', 'boolean', or 'base64'.
> > +	// Type is used by clients to provide validation of user input and direction to users.
> > +	//  Parameters used to define integer or boolean fields (e.g. replicaCount) should have the
> > +	// Type set to integer or boolean accordingly.
> > +	Type string `json:"type,omitempty"`
> > +}
> > +
> > +// TemplateParameters contains the substitution parameter overrides when processing a Template
> > +type TemplateParameters struct {
> > +	unversioned.TypeMeta `json:",inline"`
> > +
> > +	// Name is the name of the Template to be processed.
> > +	Name string `json:"name"`
>
> Looks like Brian is correct. Consider just using object meta instead, to
> be consistent with Binding.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/25293/files/58c7e4126d33836cce7082e9aa04f6818be06129..c7a56db1935dabe7aea0399ebfbed3cf5661c927#r63581936>
>

smarterclayton replied:
Never mind, that was a lie.

On Tue, May 17, 2016 at 3:43 PM, Clayton Coleman <ccoleman@redhat.com>
wrote:

> We use ObjectMeta in many places for virtual resources (like
> SubjectAccessReview).
>
> On Tue, May 17, 2016 at 2:53 PM, Daniel Smith <notifications@github.com>
> wrote:
>
>> In pkg/apis/extensions/types.go
>> <https://github.com/kubernetes/kubernetes/pull/25293#discussion_r63581936>
>> :
>>
>> > +	Required bool `json:"required,omitempty"`
>> > +
>> > +	// Type is the type that the parameter value must be parsed to. Type may be one of
>> > +	// 'string', 'integer', 'boolean', or 'base64'.
>> > +	// Type is used by clients to provide validation of user input and direction to users.
>> > +	//  Parameters used to define integer or boolean fields (e.g. replicaCount) should have the
>> > +	// Type set to integer or boolean accordingly.
>> > +	Type string `json:"type,omitempty"`
>> > +}
>> > +
>> > +// TemplateParameters contains the substitution parameter overrides when processing a Template
>> > +type TemplateParameters struct {
>> > +	unversioned.TypeMeta `json:",inline"`
>> > +
>> > +	// Name is the name of the Template to be processed.
>> > +	Name string `json:"name"`
>>
>> Looks like Brian is correct. Consider just using object meta instead, to
>> be consistent with Binding.
>>
>> —
>> You are receiving this because you commented.
>> Reply to this email directly or view it on GitHub
>> <https://github.com/kubernetes/kubernetes/pull/25293/files/58c7e4126d33836cce7082e9aa04f6818be06129..c7a56db1935dabe7aea0399ebfbed3cf5661c927#r63581936>
>>
>
>

bgrant0607 replied:
I don't want any fields here that aren't actually needed/supported.

In retrospect, ObjectMeta was confusing, since most of the fields aren't usable, and it refers to the parent resource rather than the subresource. A subset of ObjectReference probably would have been more appropriate. 

I'm open to anything that would make programmatic clients easier to write (e.g., used with the discovery API), but I don't envision users writing declarative config files for subresources.

lavalamp replied:
OK, I guess I'll live with it either way, since it's not making things
worse than they already are.

On Tue, May 17, 2016 at 10:58 PM, Brian Grant <notifications@github.com>
wrote:

> In pkg/apis/extensions/types.go
> <https://github.com/kubernetes/kubernetes/pull/25293#discussion_r63648984>
> :
>
> > +	Required bool `json:"required,omitempty"`
> > +
> > +	// Type is the type that the parameter value must be parsed to. Type may be one of
> > +	// 'string', 'integer', 'boolean', or 'base64'.
> > +	// Type is used by clients to provide validation of user input and direction to users.
> > +	//  Parameters used to define integer or boolean fields (e.g. replicaCount) should have the
> > +	// Type set to integer or boolean accordingly.
> > +	Type string `json:"type,omitempty"`
> > +}
> > +
> > +// TemplateParameters contains the substitution parameter overrides when processing a Template
> > +type TemplateParameters struct {
> > +	unversioned.TypeMeta `json:",inline"`
> > +
> > +	// Name is the name of the Template to be processed.
> > +	Name string `json:"name"`
>
> I don't want any fields here that aren't actually needed/supported.
>
> In retrospect, ObjectMeta was confusing, since most of the fields aren't
> usable, and it refers to the parent resource rather than the subresource. A
> subset of ObjectReference probably would have been more appropriate.
>
> I'm open to anything that would make programmatic clients easier to write
> (e.g., used with the discovery API), but I don't envision users writing
> declarative config files for subresources.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/25293/files/58c7e4126d33836cce7082e9aa04f6818be06129..c7a56db1935dabe7aea0399ebfbed3cf5661c927#r63648984>
>

pwittrock replied:
@bparees suggested getting rid of name and using.

`kubectl process template/<templatename> -f params.yaml`

This is cleaner in that it allows name just to be provided in one place, but isn't consistent with the rollback command.  Thoughts?

bparees replied:
@bgrant0607 @smarterclayton @lavalamp bump

bgrant0607 replied:
@bparees @pwittrock I don't understand how the kubectl proposal relates to this API discussion.

AIUI, the problem we have is that, unlike admission control, REST handlers aren't passed all of the information available from the path in the request context. If we fix that problem, then the information doesn't need to be explicitly represented in the body schema for imperative subresources that aren't serialized in manifests nor in etcd.

bparees replied:
You mean if we post the parameters to an api endpoint like:
`api/templates/<templatename>/process`

the rest handler can't determine "templatename"?

I would tend to agree that it makes more sense to fix that problem than to require the parameters struct include the templatename.  Especially since i might want to use the same parameters resource definition with multiple templates.

bgrant0607 replied:
@bparees Yes, that's exactly what I mean. I haven't looked at the code recently, but I recall it being a problem for /binding, /scale, and /rollback.

bparees replied:
@bgrant0607 definitely sounds like something we want to fix if this is the fourth endpoint to have a use for that info.

lavalamp commented on pkg/apis/extensions/types.go:
```diff
@@ -809,3 +810,80 @@ type PodSecurityPolicyList struct {
+
+	// ParameterValues is a map of substitution parameter name:value pairs to be expanded in the Template.
+	// Values defined in this map will override any defaults already set in the Template.
+	ParameterValues map[string]string `json:"parameters,omitempty"`
```
tag doesn't match field name.

lavalamp commented on pkg/apis/extensions/types.go:
```diff
@@ -809,3 +810,80 @@ type PodSecurityPolicyList struct {
+	// Name is the name of the Template to be processed.
+	Name string `json:"name"`
+
+	// ParameterValues is a map of substitution parameter name:value pairs to be expanded in the Template.
```
field name should match json tag name, not go field name.

pwittrock commented:
This should be pushed to 1.4.  Will still update with your comments @lavalamp.

ncdc commented:
@pwittrock do you still intend to try to get this in 1.4?

pwittrock commented:
@ncdc I don't think so.  Is this something that you are waiting for or were planning on?

smarterclayton commented:
I think we need to discuss in the apps sig about the timing and implementation.  It's important to us, but there's a lot of important stuff going on right now.  I think the clearest overarching theme for Kube right now is "getting started" and there are things in templates that do address that, but it'll be a balancing on other timing.

pwittrock commented:
@smarterclayton Sounds good to me.  I am hoping that Helm will be able to address some of the same things as Templates for the getting started experience.
